tests/export_test.py

```diff
@@ -1098,50 +1098,64 @@ def f_jax(x):  # x: f32[10,20] -> f32[20,10]
 
     vjp_module_str = str(exp_vjp.mlir_module())
 
+    # The MHLO attributes of the args and the result of the main function
```

Comment:
@yashk2810 This test has a surprising behavior. The argument that has a missing sharding annotation or a `None` partition spec ends up either with unspecified sharding or with replicated sharding: it has replicated sharding if we use a mesh context manager, and unspecified otherwise.

The test prints the shardings it finds in the HLO. Look how `arg0_attrs` depends on `in_shardings` (which can be `"P"` if we use a partition spec, `None`, or `"missing"` if there is no sharding annotation on the function. Whether or not with use a mesh context manager is printed as part of the test:

```
tests/export_test.py::JaxExportTest::test_grad_with_sharding_in_shardings_None_out_shardings_None_with_mesh_context_False
in_shardings=None, out_shardings=None
arg0_attrs='{mhlo.layout_mode = "default"} loc("args_and_out_cts_flat_jax[0]")'
arg1_attrs='{mhlo.layout_mode = "default"} loc("args_and_out_cts_flat_jax[1]")'
res_attrs='{jax.result_info = "[0]", mhlo.layout_mode = "default"}) {'
PASSED
tests/export_test.py::JaxExportTest::test_grad_with_sharding_in_shardings_None_out_shardings_None_with_mesh_context_True
in_shardings=None, out_shardings=None
arg0_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{replicated}"} loc("args_and_out_cts_flat_jax[0]")'
arg1_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{replicated}"} loc("args_and_out_cts_flat_jax[1]")'
res_attrs='{jax.result_info = "[0]", mhlo.layout_mode = "default", mhlo.sharding = "{replicated}"}) {'
PASSED
tests/export_test.py::JaxExportTest::test_grad_with_sharding_in_shardings_None_out_shardings_P_with_mesh_context_False
in_shardings=None, out_shardings='P'
arg0_attrs='{mhlo.layout_mode = "default"} loc("args_and_out_cts_flat_jax[0]")'
arg1_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{devices=[2,1]<=[2]}"} loc("args_and_out_cts_flat_jax[1]")'
res_attrs='{jax.result_info = "[0]", mhlo.layout_mode = "default"}) {'
PASSED
tests/export_test.py::JaxExportTest::test_grad_with_sharding_in_shardings_None_out_shardings_P_with_mesh_context_True
in_shardings=None, out_shardings='P'
arg0_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{replicated}"} loc("args_and_out_cts_flat_jax[0]")'
arg1_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{devices=[2,1]<=[2]}"} loc("args_and_out_cts_flat_jax[1]")'
res_attrs='{jax.result_info = "[0]", mhlo.layout_mode = "default", mhlo.sharding = "{replicated}"}) {'
PASSED
tests/export_test.py::JaxExportTest::test_grad_with_sharding_in_shardings_None_out_shardings_missing_with_mesh_context_False
in_shardings=None, out_shardings='missing'
arg0_attrs='{mhlo.layout_mode = "default"} loc("args_and_out_cts_flat_jax[0]")'
arg1_attrs='{mhlo.layout_mode = "default"} loc("args_and_out_cts_flat_jax[1]")'
res_attrs='{jax.result_info = "[0]", mhlo.layout_mode = "default"}) {'
PASSED
tests/export_test.py::JaxExportTest::test_grad_with_sharding_in_shardings_None_out_shardings_missing_with_mesh_context_True
in_shardings=None, out_shardings='missing'
arg0_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{replicated}"} loc("args_and_out_cts_flat_jax[0]")'
arg1_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{replicated}"} loc("args_and_out_cts_flat_jax[1]")'
res_attrs='{jax.result_info = "[0]", mhlo.layout_mode = "default", mhlo.sharding = "{replicated}"}) {'
PASSED
tests/export_test.py::JaxExportTest::test_grad_with_sharding_in_shardings_P_out_shardings_None_with_mesh_context_False
in_shardings='P', out_shardings=None
arg0_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{devices=[1,2]<=[2]}"} loc("args_and_out_cts_flat_jax[0]")'
arg1_attrs='{mhlo.layout_mode = "default"} loc("args_and_out_cts_flat_jax[1]")'
res_attrs='{jax.result_info = "[0]", mhlo.layout_mode = "default", mhlo.sharding = "{devices=[1,2]<=[2]}"}) {'
PASSED
tests/export_test.py::JaxExportTest::test_grad_with_sharding_in_shardings_P_out_shardings_None_with_mesh_context_True
in_shardings='P', out_shardings=None
arg0_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{devices=[1,2]<=[2]}"} loc("args_and_out_cts_flat_jax[0]")'
arg1_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{replicated}"} loc("args_and_out_cts_flat_jax[1]")'
res_attrs='{jax.result_info = "[0]", mhlo.layout_mode = "default", mhlo.sharding = "{devices=[1,2]<=[2]}"}) {'
PASSED
tests/export_test.py::JaxExportTest::test_grad_with_sharding_in_shardings_P_out_shardings_P_with_mesh_context_False
in_shardings='P', out_shardings='P'
arg0_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{devices=[1,2]<=[2]}"} loc("args_and_out_cts_flat_jax[0]")'
arg1_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{devices=[2,1]<=[2]}"} loc("args_and_out_cts_flat_jax[1]")'
res_attrs='{jax.result_info = "[0]", mhlo.layout_mode = "default", mhlo.sharding = "{devices=[1,2]<=[2]}"}) {'
PASSED
tests/export_test.py::JaxExportTest::test_grad_with_sharding_in_shardings_P_out_shardings_P_with_mesh_context_True
in_shardings='P', out_shardings='P'
arg0_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{devices=[1,2]<=[2]}"} loc("args_and_out_cts_flat_jax[0]")'
arg1_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{devices=[2,1]<=[2]}"} loc("args_and_out_cts_flat_jax[1]")'
res_attrs='{jax.result_info = "[0]", mhlo.layout_mode = "default", mhlo.sharding = "{devices=[1,2]<=[2]}"}) {'
PASSED
tests/export_test.py::JaxExportTest::test_grad_with_sharding_in_shardings_P_out_shardings_missing_with_mesh_context_False
in_shardings='P', out_shardings='missing'
arg0_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{devices=[1,2]<=[2]}"} loc("args_and_out_cts_flat_jax[0]")'
arg1_attrs='{mhlo.layout_mode = "default"} loc("args_and_out_cts_flat_jax[1]")'
res_attrs='{jax.result_info = "[0]", mhlo.layout_mode = "default", mhlo.sharding = "{devices=[1,2]<=[2]}"}) {'
PASSED
tests/export_test.py::JaxExportTest::test_grad_with_sharding_in_shardings_P_out_shardings_missing_with_mesh_context_True
in_shardings='P', out_shardings='missing'
arg0_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{devices=[1,2]<=[2]}"} loc("args_and_out_cts_flat_jax[0]")'
arg1_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{replicated}"} loc("args_and_out_cts_flat_jax[1]")'
res_attrs='{jax.result_info = "[0]", mhlo.layout_mode = "default", mhlo.sharding = "{devices=[1,2]<=[2]}"}) {'
PASSED
tests/export_test.py::JaxExportTest::test_grad_with_sharding_in_shardings_missing_out_shardings_None_with_mesh_context_False
in_shardings='missing', out_shardings=None
arg0_attrs='{mhlo.layout_mode = "default"} loc("args_and_out_cts_flat_jax[0]")'
arg1_attrs='{mhlo.layout_mode = "default"} loc("args_and_out_cts_flat_jax[1]")'
res_attrs='{jax.result_info = "[0]", mhlo.layout_mode = "default"}) {'
PASSED
tests/export_test.py::JaxExportTest::test_grad_with_sharding_in_shardings_missing_out_shardings_None_with_mesh_context_True
in_shardings='missing', out_shardings=None
arg0_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{replicated}"} loc("args_and_out_cts_flat_jax[0]")'
arg1_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{replicated}"} loc("args_and_out_cts_flat_jax[1]")'
res_attrs='{jax.result_info = "[0]", mhlo.layout_mode = "default", mhlo.sharding = "{replicated}"}) {'
PASSED
tests/export_test.py::JaxExportTest::test_grad_with_sharding_in_shardings_missing_out_shardings_P_with_mesh_context_False
in_shardings='missing', out_shardings='P'
arg0_attrs='{mhlo.layout_mode = "default"} loc("args_and_out_cts_flat_jax[0]")'
arg1_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{devices=[2,1]<=[2]}"} loc("args_and_out_cts_flat_jax[1]")'
res_attrs='{jax.result_info = "[0]", mhlo.layout_mode = "default"}) {'
PASSED
tests/export_test.py::JaxExportTest::test_grad_with_sharding_in_shardings_missing_out_shardings_P_with_mesh_context_True
in_shardings='missing', out_shardings='P'
arg0_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{replicated}"} loc("args_and_out_cts_flat_jax[0]")'
arg1_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{devices=[2,1]<=[2]}"} loc("args_and_out_cts_flat_jax[1]")'
res_attrs='{jax.result_info = "[0]", mhlo.layout_mode = "default", mhlo.sharding = "{replicated}"}) {'
PASSED
tests/export_test.py::JaxExportTest::test_grad_with_sharding_in_shardings_missing_out_shardings_missing_with_mesh_context_False
in_shardings='missing', out_shardings='missing'
arg0_attrs='{mhlo.layout_mode = "default"} loc("args_and_out_cts_flat_jax[0]")'
arg1_attrs='{mhlo.layout_mode = "default"} loc("args_and_out_cts_flat_jax[1]")'
res_attrs='{jax.result_info = "[0]", mhlo.layout_mode = "default"}) {'
PASSED
tests/export_test.py::JaxExportTest::test_grad_with_sharding_in_shardings_missing_out_shardings_missing_with_mesh_context_True
in_shardings='missing', out_shardings='missing'
arg0_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{replicated}"} loc("args_and_out_cts_flat_jax[0]")'
arg1_attrs='{mhlo.layout_mode = "{1,0}", mhlo.sharding = "{replicated}"} loc("args_and_out_cts_flat_jax[1]")'
res_attrs='{jax.result_info = "[0]", mhlo.layout_mode = "default", mhlo.sharding = "{replicated}"}) {'
PASSED

```

Reply:
> The argument that has a missing sharding annotation or a None partition spec ends up either with unspecified sharding or with replicated sharding: it has replicated sharding if we use a mesh context manager, and unspecified otherwise.

Yeah, that's just the difference in semantics of `jit` and `pjit`.

pjit years ago considered `None` as replicated and those semantics for `pjit` still exist today (if mesh context manager is present). But for `jit`, None means jax can choose the sharding (jit doesn't respect the context manager so the semantics are straightforward for jit).

But this should be fine right? Like export shouldn't care about it?

Reply:
export does not care about this in particular, because it will do whatever jit/pjit do, but I pointed it out because it is a strange behavior. It is strange for pjit to interpret `None` differently depending on whether there is a mesh context or not. You probably have thought about cleaning this up, can we at least add it to the list of changes to prepare for the next major version?

Reply:
Well, the way to clean it up is to delete pjit 😅

So deleting pjit should be on the list!